### ISSUE TITLE
Resetting Ubuntu VM Password does not work

### ISSUE THREAD

**Issue by nxsbi** [apache/cloudstack#8767]
##### ISSUE TYPE

 * Other

##### COMPONENT NAME

~~~
UI
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0
~~~

##### CONFIGURATION

NA

##### OS / ENVIRONMENT

Ubuntu 22.04 Server LTS (VM OS)

##### SUMMARY

I want to enable the Change Password functionality for Ubuntu based VM. I have a newly created Ubuntu 22.04 Server Template. I enabled the root user as well, and followed steps in  [https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_password.html](https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_password.html )

However, for Ubuntu I cannot get it to work.  Step 5 is not possible on Ubuntu so tried 
`update-rc.d cloud-set-guest-password defaults`

Even after trying the Change Password from Cloudstack UI, the password does not change and my old password which I used while creating template still works.

These steps worked fine for CentOS, but not working for Ubuntu

##### STEPS TO REPRODUCE

~~~
Create a Ubuntu Template
Follow the cloud-set-guest-password steps - https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_password.html
Create a new VM from this template, - New password that pops up does not work
Chnage password does not work from Cloudstack UI 

~~~

##### EXPECTED RESULTS

~~~
Password works
~~~

##### ACTUAL RESULTS

~~~
Password does not work. 
~~~

**Comment by nxsbi** [comment]
PS - I have also followed the steps in this link 

https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_cloud_init.html#linux-with-cloud-init

Created the files 
- /etc/cloud/cloud.cfg.d/99_cloudstack.cfg
- /etc/cloud/cloud.cfg.d/80_user.cfg

**Comment by kiranchavala** [comment]
@nxsbi 

Have you enabled  the  password option on the template

![Screenshot 2024-03-11 at 12 27 33 PM](https://github.com/apache/cloudstack/assets/1401014/599e223e-87d8-4e0f-8cd5-d5366979554a)

**Comment by kiranchavala** [comment]
@nxsbi 

could you please try with the following configuration 

cat /etc/cloud/cloud.cfg

```
datasource_list: ["CloudStack"]

datasource:
  CloudStack:
    max_wait: 60
    timeout: 30

network:
  config: disabled

users:
  - default

system_info:
  default_user:
    name: root

disable_root: false

ssh_deletekeys: true

cloud_init_modules:
 - seed_random
 - bootcmd
 - write-files
 - growpart
 - resizefs
 - disk_setup
 - mounts
 - set_hostname
 - update_hostname
 - update_etc_hosts
 - ssh
 - users-groups

cloud_config_modules:
 - ssh-import-id
 - locale
 - [ set-passwords, always ]
 - ntp
 - timezone
 - disable-ec2-metadata
 - runcmd

cloud_final_modules:
 - scripts-vendor
 - scripts-per-once
 - scripts-per-boot
 - scripts-per-instance
 - scripts-user
 - puppet

warnings:
  dsid_missing_source: off
```

**Comment by weizhouapache** [comment]
@nxsbi 
if you run the script `cloud-set-guest-password` manually inside the VM, is the password changed ?

**Comment by nxsbi** [comment]
@kiranchavala   Yes - the password enabled is selected. 
Also tried the config you provided, but still not working

@weizhouapache  - I tried running the cloud-set-guest-password manually, and it produces errors -- and then it hangs - never gives the shell back (have to Ctrl+C)

<img width="677" alt="image" src="https://github.com/apache/cloudstack/assets/71042351/99d4220e-f87d-45d1-a47c-02d40457a823">

**Comment by weizhouapache** [comment]
> @kiranchavala Yes - the password enabled is selected. Also tried the config you provided, but still not working
> 
> @weizhouapache - I tried running the cloud-set-guest-password manually, and it produces errors -- and then it hangs - never gives the shell back (have to Ctrl+C)
> 

@nxsbi 
try this script 

https://raw.githubusercontent.com/apache/cloudstack/main/setup/bindir/cloud-set-guest-sshkey-password-userdata-configdrive.in

**Comment by nxsbi** [comment]
@weizhouapache  Thanks - I ran the script below is the output. Also tried resetting the password from UI after that but did not work.

<img width="589" alt="image" src="https://github.com/apache/cloudstack/assets/71042351/1a97a606-2af5-499b-89e0-e2baab820818">

**Comment by weizhouapache** [comment]
> @weizhouapache Thanks - I ran the script below is the output. Also tried resetting the password from UI after that but did not work.
> 
> <img alt="image" width="589" src="https://private-user-images.githubusercontent.com/71042351/311843750-1a97a606-2af5-499b-89e0-e2baab820818.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxODY3ODUsIm5iZiI6MTcxMDE4NjQ4NSwicGF0aCI6Ii83MTA0MjM1MS8zMTE4NDM3NTAtMWE5N2E2MDYtMmFmNS00OTliLTg5ZTAtZTJiYWFiODIwODE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDE5NDgwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5MmU1NjYyZDVjMDIzYjczNjljMDBhMjBmZmUwMzhlNzBlZmVlOTJhZDE3YTFhOTg3OWNjMzliYTg5MjM5MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.z9YHrLawhPBxrp9mBHURhRH2N9E8RFwVYm51uWr3--s">

@nxsbi 
the output of the command looks ok to me.
if you change the password of a non-root user, you need to change the `username` in the script.

**Comment by nxsbi** [comment]
Yes already did that - the username was changes to correct one.  I have that working on my CentOS and Rocky Linux scripts so familair with that. 

I am walking through all steps -- Questions:
- In Networking, for CentOS, we do a if-cfg-eth0 config change. **Is there any change to do for Ubuntu?**  https://docs.cloudstack.apache.org/en/latest/adminguide/templates.html#system-preparation-for-linux 
- In LinuxOS Install section, we do a chkconfig --add cloud-set-guest-password - **Is there similar setting for Ubuntu?**
https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_password.html#adding-password-management-to-templates

**Comment by nxsbi** [comment]
I believe I found the problem and the solution. Documenting below for others for future reference.

Since my prior experience creating template revolved around CentOS/Rocky, I followed the steps shown in https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_password.html (I had this URL saved from before)

However, I determined that was not the right place for Ubuntu... 

Correct steps:
- I started here: https://docs.cloudstack.apache.org/en/latest/adminguide/templates.html#system-preparation-for-linux
- Follow the steps 1, 2 and 3 shown as Ubuntu
- For Step 4,5,6 and 7 -  Setup Cloud-init  Integration ONLY. Do not follow the "Adding Password Management to Your Templates" or "Implementing a SSH-Key bash script". 
    - @weizhouapache  NOTE  - the link on the page to Cloud-Init Integration (in step 4,5,6,7) is broken - Tried to Edit on GitHub but not confident enough to not break something else. Maybe you can review and fix the link - Correct link is: https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_cloud_init.html#linux-with-cloud-init
- Finalize the Template per Step 8, 9 and 10.


I believe I was having issues before I had followed both - "Adding Password Management" and "Cloud-Init Integration". And somehow those were conflicting. 

NOTE 2 - If you are adding Password Management, one of the step is to add the "Specify the managed user" during Cloud Init Setup and create the /etc/cloud/cloud.cfg.d/80_user.cfg. In this step, I had to modify the below line. Reason was that in the instance when I tried to do sudo command, I always kept getting error about the \". Changing to the new line fixes this for me (only tested in Ubuntu 22.04 Server LTS.)

Original 
` sudo: [\"ALL=(ALL) ALL\"]       # User permissions`
New
` sudo: [ALL=(ALL) ALL]       # User permissions`
